util.go

```diff
@@ -2,62 +2,89 @@ package relay
 
 import (
 	"encoding/binary"
-	"fmt"
+	"errors"
 	"io"
 
+	pb "github.com/libp2p/go-libp2p-circuit/pb"
+
+	ggio "github.com/gogo/protobuf/io"
+	proto "github.com/gogo/protobuf/proto"
+	peer "github.com/libp2p/go-libp2p-peer"
+	pstore "github.com/libp2p/go-libp2p-peerstore"
 	ma "github.com/multiformats/go-multiaddr"
+	mh "github.com/multiformats/go-multihash"
 )
 
-type singleByteReader struct {
-	r io.Reader
-}
+func peerToPeerInfo(p *pb.CircuitRelay_Peer) (empty pstore.PeerInfo, err error) {
+	if p == nil {
+		return empty, errors.New("nil peer")
+	}
 
-func (s *singleByteReader) ReadByte() (byte, error) {
-	var b [1]byte
-	n, err := s.r.Read(b[:])
+	h, err := mh.Cast(p.Id)
 	if err != nil {
-		return 0, err
+		return empty, err
 	}
-	if n == 0 {
-		return 0, io.ErrNoProgress
+
+	addrs := make([]ma.Multiaddr, len(p.Addrs))
+	for i := 0; i < len(addrs); i++ {
+		a, err := ma.NewMultiaddrBytes(p.Addrs[i])
+		if err != nil {
+			return empty, err
+		}
+		addrs[i] = a
 	}
 
-	return b[0], nil
+	return pstore.PeerInfo{ID: peer.ID(h), Addrs: addrs}, nil
 }
 
-func writeLpMultiaddr(w io.Writer, a ma.Multiaddr) error {
-	buf := make([]byte, binary.MaxVarintLen32+len(a.Bytes()))
-	n := binary.PutUvarint(buf, uint64(len(a.Bytes())))
-	n += copy(buf[n:], a.Bytes())
-	nw, err := w.Write(buf[:n])
-	if err != nil {
-		return err
+func peerInfoToPeer(pi pstore.PeerInfo) *pb.CircuitRelay_Peer {
+	addrs := make([][]byte, len(pi.Addrs))
+	for i := 0; i < len(addrs); i++ {
+		addrs[i] = pi.Addrs[i].Bytes()
 	}
-	if n != nw {
-		return fmt.Errorf("failed to write all bytes to writer")
-	}
-	return nil
+
+	p := new(pb.CircuitRelay_Peer)
+	p.Id = []byte(pi.ID)
+	p.Addrs = addrs
+
+	return p
 }
 
-func readLpMultiaddr(r io.Reader) (ma.Multiaddr, error) {
-	l, err := binary.ReadUvarint(&singleByteReader{r})
-	if err != nil {
-		return nil, err
-	}
+type delimitedReader struct {
+	r   io.Reader
+	buf []byte
+}
+
+// the gogo protobuf NewDelimitedReader is buffered, which may eat up stream data
```

Comment:
Well, the only slowdown is in reading the length, where it could take multiple single byte reads. 
But this only happens during handshake, so it's a drop in the water in the connection's lifetime.

I will add a comment to that effect.